parser/prism/Translator.cc

```diff
@@ -1348,12 +1348,38 @@ unique_ptr<parser::Node> Translator::patternTranslate(pm_node_t *node) {
         case PM_IN_NODE: { // An `in` pattern such as in a `case` statement, or as a standalone expression.
             auto inNode = reinterpret_cast<pm_in_node *>(node);
 
-            auto sorbetPattern = patternTranslate(inNode->pattern);
-
+            auto prismPattern = move(inNode->pattern);
             auto inlineIfSingle = true;
             auto statements = translateStatements(inNode->statements, inlineIfSingle);
 
-            return make_unique<parser::InPattern>(location, move(sorbetPattern), nullptr, move(statements));
+            if (prismPattern != nullptr &&
+                (PM_NODE_TYPE_P(prismPattern, PM_IF_NODE) || PM_NODE_TYPE_P(prismPattern, PM_UNLESS_NODE))) {
```

Comment:
I think we can remove this outer `if`/`else` check, like so:

```c++          
unique_ptr<parser::Node> sorbetGuard;

auto inlineIfSingle = true;
auto statements = translateStatements(inNode->statements, inlineIfSingle);

if (auto prismPattern = inNode->pattern; prismPattern != nullptr) {
    ENFORCE((PM_NODE_TYPE_P(prismPattern, PM_IF_NODE) || PM_NODE_TYPE_P(prismPattern, PM_UNLESS_NODE))))    
    
    ENFORCE(prismStatements->body.size == 1)

    // ... assign to `sorbetGuard` and `sorbetPattern`
}

// ...

// Always move in `sorbetGuard`. It might be `nullptr`, but that's ok.
return make_unique<parser::InPattern>(location, move(sorbetPattern), move(sorbetGuard), move(statements));
```

Reply:
`sorbetPattern` can have 3 types of results:

1. `nullptr`
2. When there are conditionals, it'd be the conditional's first (and only) statement
3. When there aren't additionals, the entire `in` node's pattern field

So I think we still need an outer if/else, which IMO is not hard to understand anyway. But I've followed other suggestions to simplify it even more 👍

Reply:
Nice!